### ISSUE TITLE
cmake-format should be optional

### ISSUE THREAD

**Issue by matz-e** [BlueBrain/hpc-coding-conventions#37]
A buggy formatting tool that still seems to be in beta status should not be required.

**Comment by tristan0x** [comment]
Can you elaborate on the issues you encounter?
CMake files should be consistent, and this tool allow such thing.

**Comment by pramodk** [comment]
Agree with @tristan0x! 

The CMake files of the project quickly becomes mess & inconsistent from my past experience. The tool like cmake-format is really helpful with experience from NMODL project.

Now, coming back to the original point of this ticket : @ohm314 is having strange issues on his desktop whereas Tristan, me and Ioannis are using cmake-format without issues. So, it would be useful to isolate and identify the actual issues (whether it's particular python version, cmake-format version etc).

**Comment by matz-e** [comment]
My main issues (I get the gist that we should have uniform formatting):

* l have to install it, and from our recent discussions, `0.4.5` and `0.5.[01]` differ in behavior. I don't think its smart to rely on non-stable software like this. @ohm314 observed some nice crashes that I have *zero* desire to reproduce.
* `clang-format` should be independently accessible from `cmake-format`.

**Comment by ohm314** [comment]
@matz-e was faster - I was gonna make this suggestion.
I disagree with @pramodk . CMakeFiles typically don't change very often. One should keep them clean of course but what I've experienced so far is that with a little care and maybe one initial manual run with a tool I spend much less time than with `cmake-format`. As @matz-e pointed out. The tool is still quite immature and undergoing big code changes and suffering from bugs (cf. release notes 0.5.0/0.5.1 https://github.com/cheshirekow/cmake_format/releases). Using tools is great, but not if they get in the way of productivity rather than enabling it. Finally, @pramodk : the reason why I was having "strange" issues, is because I changed from pyenv to system python and reinstalled dependencies pulling in first a broken version of cmake-format and now one that seems to produce a different formatting than 0.4.5.
Long story short, I agree with @matz-e 's request :)

**Comment by pramodk** [comment]
> One should keep them clean of course but what I've experienced so far is that with a little care and maybe one initial manual run with a tool I spend much less time than with cmake-format.

Oh yes..this seems very familiar...especially "one initial manual run"... I disagree ;)

**Comment by pramodk** [comment]
>  cmake-format and now one that seems to produce a different formatting than 0.4.5

@ohm314  : just curious, does 0.4.5 with system python 3.x works? Wondering if I will also hit this issue and start complaining about the same :)

**Comment by matz-e** [comment]
> >   One should keep them clean of course but what I've experienced so far is that with a little care and maybe one initial manual run with a tool I spend much less time than with cmake-format.
>
> Oh yes..this seems very familiar...especially "one initial manual run"... I disagree ;)

So we have go-ahead to use beta-quality software? I'll quote you on that :)

**Comment by pramodk** [comment]
Well...I am not saying `cmake-format` is `the` solution.. I have very little idea about it anyway.

Obviously if there are better alternatives, we should go for it. But if not, can we identify / isolate issue and see if that is solvable/avoidable in our scenario? For example, not trying to defend cmake-format but : 

> * and from our recent discussions, 0.4.5 and 0.5.[01] differ in behavior. 

It's same for clang-format...from version to version, the formatting change even if we use same `.clang-format` file (not comparing clang-format for stability here!)

> * The project does not mention its status (no alpha/beta/stable indicator in setup.py)

I assume this is the case with many python packages :) 

> clang-format should be independently accessible from cmake-format

Agree...if I am not mistaken, currently we have single option to enable clang-format and cmake-format. There could be separate option for those.

> is because I changed from pyenv to system python and reinstalled dependencies pulling in first a broken version of cmake-format and now one that seems to produce a different formatting than 0.4.5.

that's what I wanted to understand better...Is that `0.4.5` with particular python version is an issue? Because I have been using following without issue: 

```
→ pip list | grep cmake-format
cmake-format                  0.4.5
→ python --version
Python 3.7.2
```

**Comment by tristan0x** [comment]
> Using tools is great, but not if they get in the way of productivity rather than enabling it

Would it be possible to elaborate in which way it decrease your productivity? I am not sure to understand. From the usage perspective, you just have to execute `make clang-format`. If it refers to the setup issue you mentioned above, I agree it is a pity you installed a malfunctioning version. But this issue can be easily fixed by forcing the desired version in the `find_package(CMAKE_FORMAT_VERSION 1.2.3)` in hpc-coding-conventions.
I understand it can be difficult to manage several python environments. If you are not comfortable with it, you can maybe rely on spack setup/build-env/dyi to setup the python environment, see [a smart suggestion from @pramodk ](https://github.com/BlueBrain/spack/blob/develop/var/spack/repos/builtin/packages/zee/package.py#L32-L35)

Last but not least, did you have the chance to provide feedback to cmake-format about the crashes you experienced? (it is almost 1am here, my apologies for not checking the cmake-format issue) tracker)

Have a good week-end all.

**Comment by ohm314** [comment]
No worries @tristan0x , @pramodk I really don't want to drag this into an endless discussion...
To be clear this is not a critique at all of @tristan0x 's excellent work on hpc-coding-conventions and I strongly support the effort of improving our coding standards, both in the code itself and the build system. But I found myself particularly with the `cmake-format` issues blocked or at least distracted from working on the actual project/issue/pr I wanted to. Of course, all the problems that we are discussing here can be solved. The point is that I think we should not be using tools and processes which we have to debug and fix instead of working on more interesting projects. I personally found it quite aggravating to run into these issues at the moment when I wanted to commit and push my code. 
Also, note that @matz-e and I have singled out `cmake-format` because it is causing trouble. We're not saying we should drop everything and go back to our old ways. `clang-format` is great and I wouldn't want to skip that.
About the different versions and python versions I don't know @pramodk . I suspect the combination of python and `cmake-format` doesn't matter. It seems a problem that popped up with the latest versions of `cmake-format`.

**Comment by tristan0x** [comment]
Thank you for your kind message @ohm314 
This hpc-coding-conventions repository is supposed to improve code and save developer's time. So for sure, regarding your recent experience, it requires improvements.
 
In the future, if you come across similar pre-commits issues and just want to get things done, you can use the following `git commit` option:
```
    -n, --no-verify
           This option bypasses the pre-commit and commit-msg hooks. See also githooks(5).
```

This is just a workaround of issues that should be reported and fixed at some point. But at least it will allow you to keep working on things that matter most.

**Comment by ohm314** [comment]
Thanks, @tristan0x :) I'm way ahead of you, that option may have come to use on a few commits ;)

Anyway, I will create a PR of a few changes that I've tried yesterday to implement in this repo.

**Comment by matz-e** [comment]
> Well...I am not saying `cmake-format` is `the` solution.. I have very little idea about it anyway.

It seems to be the only tool to format CMake files, but the developmental habits of it seems to go against development practices we try to strive for :-/

> > clang-format should be independently accessible from cmake-format
> 
> Agree...if I am not mistaken, currently we have single option to enable clang-format and cmake-format. There could be separate option for those.

I'll maybe take a stab at that. This is what I also meant by being *optional* (`PYTHON_EXECUTABLE` should also not be needed if I want just `clang-format`).

**Comment by tristan0x** [comment]
> I'll maybe take a stab at that. This is what I also meant by being _optional_ (`PYTHON_EXECUTABLE` should also not be needed if I want just `clang-format`).

On the contrary!
Using Python scripts was the fastest way for me to do it, but I would love to review a pull-request that get rid of this dependency without losing the current features. This could be done with 100% CMake code and scripts for instance, but I doubt it would be more understandable and maintainable than the current code.

That being said, IMO Python 3.5 or higher is not too much to ask, and it offers a lot of flexibility in the project.

**Comment by matz-e** [comment]
> > I'll maybe take a stab at that. This is what I also meant by being _optional_ (`PYTHON_EXECUTABLE` should also not be needed if I want just `clang-format`).
> 
> On the contrary!
> Using Python scripts was the fastest way for me to do it, but I would love to review a pull-request that get rid of this dependency without losing the current features. This could be done with 100% CMake code and scripts for instance, but I doubt it would be more understandable and maintainable than the current code.
> 
> That being said, IMO Python 3.5 or higher is not too much to ask, and it offers a lot of flexibility in the project.

So are you saying that all of the formatting targets (not the hooks) will require Python?
